iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -2106,6 +2106,7 @@ static void process_queued_ack_messages(PMQTTTRANSPORT_HANDLE_DATA transport_dat
                 free(msg_detail_entry);
 
                 DisconnectFromClient(transport_data);
+                transport_data->transport_callbacks.connection_status_cb(IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, IOTHUB_CLIENT_CONNECTION_RETRY_EXPIRED, transport_data->transport_ctx);
```

Comment:
Curious if you can double check something here?

Looking at this, this gets called here
https://github.com/Azure/azure-iot-sdk-c/blob/26d8aed1362e4434afcf2f62267995bb92e6c262/iothub_client/src/iothubtransport_mqtt_common.c#L3272 in `IoTHubTransport_MQTT_Common_DoWork()`.

If [`InitializeConnection()`](https://github.com/Azure/azure-iot-sdk-c/blob/26d8aed1362e4434afcf2f62267995bb92e6c262/iothub_client/src/iothubtransport_mqtt_common.c#L3176) fails, this section might not have been hit:
https://github.com/Azure/azure-iot-sdk-c/blob/26d8aed1362e4434afcf2f62267995bb92e6c262/iothub_client/src/iothubtransport_mqtt_common.c#L2434-L2438
which sets the callback and sets that "set" flag to true. I'm wondering if this line you added should be wrapped in a similar `if (!transport_data->isRetryExpiredCallbackSet)` to prevent invoking a callback which is `NULL`.

Hate to be that person but there's a chance there should be a UT here for this scenario if that's true.

Reply:
No you're good, thank you for your attention to detail here.  Especially since this space is new for me.  What is a UT?  And what do you mean by patching up the expected calls ut stuff?

Reply:
Unit tests. So if you add a new function call in one of our API's, you likely have to change the UT's since they have a list of "expected calls" to verify. The idea being that with a given UT, you expect the code to flow a certain, defined way. We set up "expected calls" to verify that indeed, the flow you expected was called. So in this case, there's a chance that a new call breaks the previous expected "flow".

Reply:
example I think in the pipeline here:
https://azure-iot-sdks.visualstudio.com/azure-iot-sdks/_build/results?buildId=72109&view=logs&j=d89e688f-0939-5d98-042b-db98508a1b62&t=8fca59f4-1111-5c89-dc70-3102283020d1&l=1755

Reply:
awesome, thank you!

Reply:
Q:  How does making the call `transport_data->transport_callbacks.connection_status_cb(...` set the callback?  I just see the callback called, not "set".